File: 3427-special-array-ii/3427-special-array-ii.java
class Solution {
    public boolean[] isArraySpecial(int[] nums, int[][] queries) {
        int n = nums.length;
        boolean prev = (nums[0] % 2 == 1);
        List<Integer> index = new ArrayList<>();
        for (int i = 1; i < n; i++) {
            boolean curr = (nums[i] % 2 == 1);
            if (prev == curr) {
                index.add(i - 1);
            }
            prev = curr;
        }
        boolean[] ans = new boolean[queries.length];
        for (int q = 0; q < queries.length; q++) {
            int from = queries[q][0], to = queries[q][1];
            int left = 0, right = index.size() - 1, found = -1;
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (index.get(mid) >= from) {
                    found = mid;
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            ans[q] = (found == -1 || to <= index.get(found));
        }
        return ans;
    }
}

File: 3427-special-array-ii/3427-special-array-ii.js
/**
 * @param {number[]} nums
 * @param {number[][]} queries
 * @return {boolean[]}
 */
var isArraySpecial = function(nums, queries) {
    const n = nums.length;
    const index = [];
    let prev = nums[0] % 2 === 1;

    for (let i = 1; i < n; i++) {
        const curr = nums[i] % 2 === 1;
        if (prev === curr) {
            index.push(i - 1);
        }
        prev = curr;
    }

    const result = [];
    for (const [from, to] of queries) {
        let left = 0, right = index.length - 1, found = -1;
        while (left <= right) {
            const mid = Math.floor((left + right) / 2);
            if (index[mid] >= from) {
                found = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        result.push(found === -1 || to <= index[found]);
    }

    return result;
};
